relentless/optimize/method.py

```diff
@@ -177,41 +193,67 @@ def __init__(self, abs_tol, max_iter):
         self.abs_tol = abs_tol
         self.max_iter = max_iter
 
-    def find(self, objective, start, end):
+    def find(self, objective, start, end, grad_scale=1.0):
         """Apply the line search algorithm to take the optimal step.
 
         Note that the objective function is kept at its initial state, and the
         function evaluted after taking the optimal step is returned separately.
 
         Parameters
         ----------
-        objective : :class:`~relentless.optimize.ObjectiveFunction`
+        objective : :class:`~relentless.optimize.objective.ObjectiveFunction`
             The objective function for which the line search is applied.
-        start : :class:`~relentless.optimize.ObjectiveFunctionResult`
+        start : :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
             The objective function evaluated at the start of the search interval.
-        end : :class:`~relentless.optimize.ObjectiveFunctionResult`
+        end : :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
             The objective function evaluated at the end of the search interval.
+        grad_scale : float or dict or :class:`~relentless._collections.KeyedArray`
+            The gradient scaling parameter or parameters (:math:`\mathbf{b}`)
+            keyed on the :class:`~relentless.optimize.objective.ObjectiveFunction`
+            design variables (defaults to ``1.0``).
 
         Raises
         ------
+        ValueError
+            If the gradient scaling parameters are not positive.
         ValueError
             If the defined search interval is not a descent direction.
 
         Returns
         -------
-        :class:`~relentless.optimize.ObjectiveFunctionResult`
-            The objective function evaluated at the new, 'optimal' location.
+        :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
+            The objective function evaluated at the new, "optimal" location.
 
         """
+        # check if valid scaling
+        if isinstance(grad_scale, _collections.KeyedArray):
+            scale = grad_scale
+            err = any([sc <= 0 for sc in grad_scale.todict().values()])
+        else:
+            try:
+                scale = dict(grad_scale)
+                scale = _collections.KeyedArray(keys=grad_scale.keys())
+                scale.update(grad_scale)
+                err = any([sc <= 0  for sc in grad_scale.values()])
+            except TypeError:
+                scale = grad_scale
+                err = grad_scale <= 0
+        if err:
+            raise ValueError('The gradient scaling parameters must be positive.')
+
         ovars = {x: x.value for x in objective.design_variables()}
 
         # compute normalized search direction
         d = end.design_variables - start.design_variables
         max_step = d.norm()
+        if max_step == 0:
```

Comment:
The case you're guarding against should return exactly equal, but I would use `isclose` since this is a floating-point value.
```suggestion
        if np.isclose(max_step,0):
```